File: app/src/main/java/cn/xhuww/mvvm/base/BaseFragment.java
package cn.xhuww.mvvm.base;

import android.content.Context;
import android.databinding.DataBindingUtil;
import android.databinding.ViewDataBinding;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentActivity;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;
import android.widget.Toast;

import javax.inject.Inject;

import dagger.android.support.AndroidSupportInjection;
import dagger.android.support.DaggerFragment;

/**
 * Created by xhu_ww on 2018/5/9.
 * description: V4包下的Fragment
 */
public abstract class BaseFragment<V extends BaseViewModel> extends Fragment {

    protected View mRootView;
    protected FragmentActivity activity;
    protected final String TAG = getClass().getName();

    @Inject
    public V viewModel;

    @Override
    public void onAttach(Context context) {
        activity = (FragmentActivity) context;
        super.onAttach(context);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        mRootView = LayoutInflater.from(getContext()).inflate(getLayoutRes(), container, false);
        return mRootView;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        Log.i(TAG, "------onCreateView");
        initData();
        initView(view);
        httpRequest();
    }

    protected abstract int getLayoutRes();

    /**
     * 初始化数据，变量这些，执行位置在 initView之前
     */
    protected abstract void initData();

    protected abstract void initView(View view);

    /**
     * 网络请求
     */
    protected abstract void httpRequest();

    protected void toast(String content) {
        activity.runOnUiThread(() -> Toast.makeText(activity, content, Toast.LENGTH_SHORT).show());
    }


    @Override
    public void onResume() {
        super.onResume();
        Log.i(TAG, "------onResume");
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        Log.i(TAG, "------onDestroy");
    }

    @Override
    public void onHiddenChanged(boolean hidden) {
        super.onHiddenChanged(hidden);
        Log.i(TAG, "------onHiddenChanged:" + hidden);
    }
}

File: kotlin_mvvm/src/main/java/com/example/study/base/BaseFragment.kt
package com.example.study.base

import android.databinding.DataBindingUtil
import android.databinding.ViewDataBinding
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import dagger.android.support.DaggerFragment
import javax.inject.Inject

/**
 * Created by xhu_ww on 2018/5/29.
 * description: support 下的Fragment
 *
 */
abstract class BaseFragment<B : ViewDataBinding, V : BaseViewModel> : DaggerFragment() {

    lateinit var binding: B
    val TAG = this.javaClass.name

    @Inject
    lateinit var viewModel: V

    protected lateinit var mRootView: View
    /**
     * 布局ID
     */
    abstract fun getLayoutId(): Int

    /**
     * 初始化数据 注：init 在 initView前
     */
    abstract fun init()

    /**
     * 初始化View
     */
    abstract fun initView(rootView: View)

    /**
     * 展示数据
     */
    abstract fun showData()

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        binding = DataBindingUtil.inflate(inflater, getLayoutId(), container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        init()
        initView(view)
        showData()
        Log.i(TAG, "---------onViewCreated")
    }

    fun toast(message: String) {
        Toast.makeText(context, message, Toast.LENGTH_SHORT).show()
    }

    override fun onDestroy() {
        viewModel.dispose()
        Log.i(TAG, "---------onDestroy")
        super.onDestroy()
    }

    override fun onResume() {
        super.onResume()
        Log.i(TAG, "---------onResume")
    }

}